addons/mail/tests/test_ir_ui_menu.py

```diff
@@ -0,0 +1,94 @@
+                'view_mode': view_mode,
+                'domain': domain,
+                'context': context or {},
+                'type': 'ir.actions.act_window',
+            })
+
+        # Remove all menus and setup test menu with known results
+        Menu.search([]).unlink()
+
+        menu_root_invoicing = new_menu('Invoicing')
+        menu_invoicing_customer = new_menu('Customers', parent_id=menu_root_invoicing.id)
+        new_menu('Customers', parent_id=menu_invoicing_customer.id,
+                 action=new_action('Customers', res_model='res.partner', view_mode='kanban,tree,form',
+                                   context="""{
+                                               'search_default_customer': 1,
+                                               'res_partner_search_mode': 'customer', 
+                                               'default_is_company': True, 
+                                               'default_customer_rank': 1
+                                           }"""))
+        menu_root_contact = new_menu('Contacts')
+        new_menu('Contacts', parent_id=menu_root_contact.id,
+                 action=new_action('Contacts', res_model='res.partner', view_mode='kanban,tree,form,activity',
+                                   context="{'default_is_company': " + (' ' * 1000) + "True" + ('\n' * 1000) + "}",
+                                   domain='[]'))
+        menu_root_sales = new_menu('Sales')
+        menu_sales_orders = new_menu('Orders', parent_id=menu_root_sales.id)
+        new_menu('Customers', parent_id=menu_sales_orders.id,
+                 action=new_action('Customers', res_model='res.partner', view_mode='kanban,tree,form',
+                                   context="""{
+                                                    'search_default_customer': 1,
+                                                    'res_partner_search_mode': 'customer', 
+                                                    'default_is_company': True, 
+                                                    'default_customer_rank': 1
+                                                }"""))
+        menu_root_settings = new_menu('Settings')
+        menu_settings_user_and_companies = new_menu('Users & Companies', parent_id=menu_root_settings.id)
+        new_menu('Companies', parent_id=menu_settings_user_and_companies.id,
+                 action=new_action('Companies', res_model='res.company', view_mode='tree,kanban,form'))
+
+        cls.menu_count = new_menu.call_count
+        cls.menu_root_contact = menu_root_contact
+        cls.menu_root_settings = menu_root_settings
+        cls.Menu = Menu
+
+    def test_menu_setup(self):
+        self.assertEqual(len(self.Menu._visible_menu_ids()), self.menu_count)
+
+    def test_look_for_existing_menu_root(self):
+        self.env['ir.ui.menu']._visible_menu_ids()
+        self.env.invalidate_all()
+        with self.assertQueryCount(__system__=3):
```

Comment:
You can use the "warmup" decorator for query counter tests, that way the cache is warmed. And normally you won't have to do the "visible menu ids" trick anymore.